test_large_zip fails in i586 after #15893

Hi, it seems #15893 is making the test_large_zip test fail in openSUSE's build service for the i586 architecture with openSUSE Tumbleweed. The test output is:

```
______________________________________________________________________________________ TestSaveTxt.test_large_zip ______________________________________________________________________________________
[gw1] linux -- Python 3.8.3 /usr/bin/python3

self = <numpy.lib.tests.test_io.TestSaveTxt object at 0xe935e610>

    @pytest.mark.skipif(sys.platform=='win32', reason="files>4GB may not work")
    @pytest.mark.slow
    @requires_memory(free_bytes=7e9)
    def test_large_zip(self):
        def check_large_zip():
            # The test takes at least 6GB of memory, writes a file larger than 4GB
            test_data = np.asarray([np.random.rand(np.random.randint(50,100),4)
                                   for i in range(800000)], dtype=object)
            with tempdir() as tmpdir:
                np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
        # run in a subprocess to ensure memory is released on PyPy, see gh-15775
        p = Process(target=check_large_zip)
        p.start()
        p.join()
>       assert p.exitcode == 0
E       assert 1 == 0
E         +1
E         -0

../../../BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/tests/test_io.py:587: AssertionError
----------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/abuild/rpmbuild/BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/tests/test_io.py", line 582, in check_large_zip
    np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
  File "<__array_function__ internals>", line 5, in savez
  File "/home/abuild/rpmbuild/BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/npyio.py", line 616, in savez
    _savez(file, args, kwds, False)
  File "/home/abuild/rpmbuild/BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/npyio.py", line 721, in _savez
    format.write_array(fid, val,
  File "/home/abuild/rpmbuild/BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/format.py", line 664, in write_array
    pickle.dump(array, fp, protocol=3, **pickle_kwargs)
MemoryError
======================================================================================= short test summary info ========================================================================================
FAILED ../../../BUILDROOT/python-numpy-1.19.0-0.i386/usr/lib/python3.8/site-packages/numpy/lib/tests/test_io.py::TestSaveTxt::test_large_zip - assert 1 == 0
```

It seems other architectures aren't affected. I tried calling the test without using a separate process and that works fine for i586.

So applying [this patch](https://build.opensuse.org/package/view_file/home:alarrosa:branches:devel:languages:python:numeric/python-numpy/fix-test_large_zip-in-i586.patch?expand=1) _fixes_ it, but I didn't submit it as a PR because I understand that that would be basically reverting #15893 for i586 and that would bring #15775 back.

### Reproducing code example:

Just running the tests in a i586 architecture should show the error.

I understand that setting up such a system might not be quick, so since I can reproduce it easily, if there's any test that would help, please, tell me.

Also, I can provide help to set up a local checkout to test running the tests exactly as they're ran on the obs servers. It's basically a matter of creating an account at https://build.opensuse.org, installing the osc tool and running something like:

```
osc branch devel:languages:python:numeric python-numpy
osc checkout home:<your_obs_username>:branches:devel:languages:python:numeric python-numpy
cd home:<your_obs_username>:branches:devel:languages:python:numeric/python-numpy
osc build openSUSE_Tumbleweed i586
```


### Numpy/Python version information:

1.19.0 3.8.3 (default, May 17 2020, 14:48:56) [GCC]

--- a/numpy/lib/tests/test_io.py
+++ b/numpy/lib/tests/test_io.py
@@ -13,7 +13,8 @@ from tempfile import NamedTemporaryFile
 from io import BytesIO, StringIO
 from datetime import datetime
 import locale
-from multiprocessing import Process
+from multiprocessing import Process, Value
+from ctypes import c_bool
 
 import numpy as np
 import numpy.ma as ma
@@ -574,16 +575,26 @@ class TestSaveTxt:
     @pytest.mark.slow
     @requires_memory(free_bytes=7e9)
     def test_large_zip(self):
-        def check_large_zip():
+        def check_large_zip(memoryerror_raised):
+            memoryerror_raised.value=False
             # The test takes at least 6GB of memory, writes a file larger than 4GB
             test_data = np.asarray([np.random.rand(np.random.randint(50,100),4)
                                    for i in range(800000)], dtype=object)
             with tempdir() as tmpdir:
-                np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
+                try:
+                    np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
+                except MemoryError:
+                    memoryerror_raised.value=True
+                    raise
         # run in a subprocess to ensure memory is released on PyPy, see gh-15775
-        p = Process(target=check_large_zip)
+        # Use an object in shared memory to re-raise the MemoryError exception
+        # in our process if needed, see gh-16889
+        memoryerror_raised = Value(c_bool)
+        p = Process(target=check_large_zip, args=(memoryerror_raised,))
         p.start()
         p.join()
+        if memoryerror_raised.value:
+            raise MemoryError("Child process raised a MemoryError exception")
         assert p.exitcode == 0
 
 class LoadTxtBase:

--- a/numpy/lib/tests/test_io.py
+++ b/numpy/lib/tests/test_io.py
@@ -577,15 +577,18 @@ class TestSaveTxt:
     def test_large_zip(self):
         def check_large_zip(memoryerror_raised):
             memoryerror_raised.value = False
-            # The test takes at least 6GB of memory, writes a file larger than 4GB
-            test_data = np.asarray([np.random.rand(np.random.randint(50,100),4)
-                                   for i in range(800000)], dtype=object)
-            with tempdir() as tmpdir:
-                try:
-                    np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
-                except MemoryError:
-                    memoryerror_raised.value = True
-                    raise
+            try:
+                # The test takes at least 6GB of memory, writes a file larger
+                # than 4GB
+                test_data = np.asarray([np.random.rand(
+                                        np.random.randint(50,100),4)
+                                        for i in range(800000)], dtype=object)
+                with tempdir() as tmpdir:
+                    np.savez(os.path.join(tmpdir, 'test.npz'),
+                             test_data=test_data)
+            except MemoryError:
+                memoryerror_raised.value = True
+                raise
         # run in a subprocess to ensure memory is released on PyPy, see gh-15775
         # Use an object in shared memory to re-raise the MemoryError exception
         # in our process if needed, see gh-16889

--- a/numpy/lib/tests/test_io.py
+++ b/numpy/lib/tests/test_io.py
@@ -576,7 +576,7 @@ class TestSaveTxt:
     @requires_memory(free_bytes=7e9)
     def test_large_zip(self):
         def check_large_zip(memoryerror_raised):
-            memoryerror_raised.value=False
+            memoryerror_raised.value = False
             # The test takes at least 6GB of memory, writes a file larger than 4GB
             test_data = np.asarray([np.random.rand(np.random.randint(50,100),4)
                                    for i in range(800000)], dtype=object)
@@ -584,7 +584,7 @@ class TestSaveTxt:
                 try:
                     np.savez(os.path.join(tmpdir, 'test.npz'), test_data=test_data)
                 except MemoryError:
-                    memoryerror_raised.value=True
+                    memoryerror_raised.value = True
                     raise
         # run in a subprocess to ensure memory is released on PyPy, see gh-15775
         # Use an object in shared memory to re-raise the MemoryError exception

